### ISSUE TITLE
Inappropriate hanging indent introduced by L003

### ISSUE THREAD

**Issue by seve-martinez** [sqlfluff/sqlfluff#3836]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

`WHEN` conditions within `CASE` statements get dedented to be inline with the `CASE`.

I've tried configuring handing indents but that doesn't seem to resolve the problem. 

```
select
1,
case 
when 1 = 1
then 2
when 2 = 1
then 3 end as foo
from bar
```

the

### Expected Behaviour

Given the following query

```
SELECT
 foo
,  CASE
 WHEN 1 = 1
 THEN 2
 END AS example
FROM tbl
```

The query should return

```
SELECT
    foo
    ,  CASE
           WHEN 1 = 1
           THEN 2
    END AS example
FROM tbl
```

### Observed Behaviour

```
SELECT
    foo
    ,  CASE
       WHEN 1 = 1
          THEN 2
    END AS example
FROM tbl
```

### How to reproduce
```
import textwrap
from sqlfluff.core import Linter
from sqlfluff.core.config import FluffConfig

linter = Linter(
    config=FluffConfig(
        configs={
            "rules": {
                "comma_style": "leading",
                "tab_space_size": 2,
                "L003": {"hanging_indents": True, "indent_unit": "space", "tab_space_size": 2},
            },
        }
    )
)
```

```
linter.lint_string_wrapped(
    textwrap.dedent(
        """\
        SELECT
          foo
          , case when 1 = 1 then 2 end AS example
        FROM tbl
        """
    )
).as_records()
```

### Dialect

snowflake

### Version

1.2.1

### Configuration
```
[sqlfluff]
# verbose is an integer (0-2) indicating the level of log output
verbose = 0
# Turn off color formatting of output
nocolor = False
# Supported dialects https://docs.sqlfluff.com/en/stable/dialects.html
# Or run 'sqlfluff dialects'
dialect = snowflake
# One of [raw|jinja|python|placeholder]
templater = jinja
# Comma separated list of rules to check, default to all
rules = all
# Comma separated list of rules to exclude, or None
exclude_rules = L031,L057,L051,L060
# The depth to recursively parse to (0 for unlimited)
recurse = 0
# Below controls SQLFluff output, see max_line_length for SQL output
output_line_length = 80
# Number of passes to run before admitting defeat
runaway_limit = 10
# Ignore errors by category (one or more of the following, separated by commas: lexing,linting,parsing,templating)
ignore = None
# Ignore linting errors found within sections of code coming directly from
# templated code (e.g. from within Jinja curly braces. Note that it does not
# ignore errors from literal code found within template loops.
ignore_templated_areas = True
# can either be autodetect or a valid encoding e.g. utf-8, utf-8-sig
encoding = autodetect
# Ignore inline overrides (e.g. to test if still required)
disable_noqa = False
# Comma separated list of file extensions to lint
# NB: This config will only apply in the root folder
sql_file_exts = .sql,.sql.j2,.dml,.ddl
# Allow fix to run on files, even if they contain parsing errors
# Note altering this is NOT RECOMMENDED as can corrupt SQL
fix_even_unparsable = False
# Very large files can make the parser effectively hang.
# This limit skips files over a certain character length
# and warns the user what has happened.
# Set this to 0 to disable.
large_file_skip_char_limit = 20000

[sqlfluff:indentation]
# See https://docs.sqlfluff.com/en/stable/indentation.html
indented_joins = False
indented_ctes = False
indented_using_on = True
indented_on_contents = True
template_blocks_indent = True

[sqlfluff:templater]
unwrap_wrapped_queries = True

[sqlfluff:templater:jinja]
apply_dbt_builtins = True

[sqlfluff:templater:dbt]
project_dir = .
profiles_dir = .
profile = snowflake-dbt
target = prod

[sqlfluff:templater:jinja:macros]
# Macros provided as builtins for dbt projects
dbt_ref = {% macro ref(model_ref) %}{{model_ref}}{% endmacro %}
dbt_source = {% macro source(source_name, table) %}{{source_name}}_{{table}}{% endmacro %}
dbt_config = {% macro config() %}{% for k in kwargs %}{% endfor %}{% endmacro %}
dbt_var = {% macro var(variable, default='') %}item{% endmacro %}
dbt_is_incremental = {% macro is_incremental() %}True{% endmacro %}


[sqlfluff:templater:dbt:context]
# Put vars defined in dbt_project.yml here


[sqlfluff:templater:jinja:context]
# Put standard jinja params here

# Some rules can be configured directly from the config common to other rules
[sqlfluff:rules]
tab_space_size = 2
max_line_length = 80
indent_unit = space
comma_style = leading
allow_scalar = True
single_table_references = consistent
unquoted_identifiers_policy = all

# Some rules have their own specific config
[sqlfluff:rules:L003]
hanging_indents = True

[sqlfluff:rules:L007]
operator_new_lines = after

[sqlfluff:rules:L010]
# Keywords
capitalisation_policy = upper
# Comma separated list of words to ignore for this rule
ignore_words = None
ignore_words_regex = None

[sqlfluff:rules:L011]
# Aliasing preference for tables
aliasing = explicit

[sqlfluff:rules:L012]
# Aliasing preference for columns
aliasing = explicit

[sqlfluff:rules:L014]
# Unquoted identifiers
extended_capitalisation_policy = lower
# Comma separated list of words to ignore for this rule
ignore_words = None
ignore_words_regex = None

[sqlfluff:rules:L016]
# Line length
ignore_comment_lines = False
ignore_comment_clauses = False

[sqlfluff:rules:L027]
# Comma separated list of words to ignore for this rule
ignore_words = None
ignore_words_regex = None

[sqlfluff:rules:L026]
# References must be in FROM clause
# Disabled for some dialects (e.g. bigquery)
force_enable = False

[sqlfluff:rules:L028]
force_enable = False

[sqlfluff:rules:L029]
unquoted_identifiers_policy = aliases
quoted_identifiers_policy = none
ignore_words = None
ignore_words_regex = None

[sqlfluff:rules:L030]
extended_capitalisation_policy = upper
ignore_words = None
ignore_words_regex = None

[sqlfluff:rules:L031]
force_enable = False

[sqlfluff:rules:L036]
wildcard_policy = single

[sqlfluff:rules:L038]
select_clause_trailing_comma = forbid

[sqlfluff:rules:L040]
capitalisation_policy = upper
ignore_words = None
ignore_words_regex = None

[sqlfluff:rules:L042]
forbid_subquery_in = join

[sqlfluff:rules:L047]
prefer_count_1 = False
prefer_count_0 = False

[sqlfluff:rules:L051]
fully_qualify_join_types = inner

[sqlfluff:rules:L052]
multiline_newline = False
require_final_semicolon = False

[sqlfluff:rules:L054]
group_by_and_order_by_style = consistent

[sqlfluff:rules:L057]
unquoted_identifiers_policy = all
quoted_identifiers_policy = all
allow_space_in_identifier = False
additional_allowed_characters = ""
ignore_words = None
ignore_words_regex = None

[sqlfluff:rules:L059]
prefer_quoted_identifiers = False
ignore_words = None
ignore_words_regex = None
force_enable = False

[sqlfluff:rules:L062]
blocked_words = None
blocked_regex = None

[sqlfluff:rules:L063]
extended_capitalisation_policy = consistent
ignore_words = None
ignore_words_regex = None

[sqlfluff:rules:L064]
preferred_quoted_literal_style = consistent
force_enable = False

[sqlfluff:rules:L066]
min_alias_length = None
max_alias_length = None
```
### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by alanmcruickshank** [comment]
I think I agree with your suggested indentation, the only thing I'm not sure about is the interaction with leading commas.

In a trailing comma world, then I think we agree it should look like this:

```sql
SELECT
    some_column,
    CASE
        WHEN 1
            THEN "foo"
        ELSE "bar"
    END as other_column
FROM tbl
```

What I'm not sure about is whether in a leading comma world it should be:

```sql
SELECT
    some_column
    , CASE
        WHEN 1
            THEN "foo"
        ELSE "bar"
    END as other_column
FROM tbl
```

or what I think you're implying is (using 4 space tabs not two space tabs):

```sql
SELECT
    some_column
    , CASE
            WHEN 1
                THEN "foo"
            ELSE "bar"
        END as other_column
FROM tbl
```

I think the issue here is that `, ` take up the same space as `  ` (two spaces) and that means when we indent one step from the comma `,`, then we end up in line with the `CASE` which _looks wrong_.

Can you clarify what you'd expect?

I'm going to be in this code a lot at the moment so am happy to take a go at sorting it out, whatever we agree.

**Comment by alanmcruickshank** [comment]
To be clear, I think with `tab_space_size` set to 2 and leading commas, the current _correct_ indentation should be as below (as per the current settings - happy to discuss whether it should be different). I suspect the _bug_ here is that it's treating the `WHEN` as a hanging indent from `CASE` (which I don't think is the desired behaviour), and (although a different rule would enforce it), there's excess leading whitespace before `CASE` (with two spaces), and the `THEN` has been _over indented_ with a double indent with respect to `CASE`.

```sql
SELECT
  foo
  , CASE
    WHEN 1 = 1
      THEN 2
  END AS example
FROM tbl
```

**Comment by seve-martinez** [comment]
Thanks for the response. I wasn't sure if it was necessarily a _bug_ but yeah I think we're on the same page. 

Just so I understand, the indentation count starts from the comma, so when we indent `WHEN` it ends up in line with the `CASE` when using 2 space tabs. The code doesn't _start_ the indent from where ever the `CASE` is itself.

**Comment by alanmcruickshank** [comment]
Correct, indents are always relative to the _start_ of the previous line.
